pubtools/pulplib/_impl/model/repository/file.py

```diff
@@ -20,3 +23,53 @@ class FileRepository(Repository):
     )
 
     mutable_urls = attr.ib(default=attr.Factory(lambda: ["PULP_MANIFEST"]), type=list)
+
+    def upload_file(self, filename, relative_url=None):
+        """Upload a file to this repository.
+
+        The upload operation includes 4 steps:
+            1. Request an upload id from pulp
+            2. With the requested upload id, upload the file
+            3. After the upload's done, import the uploaded file to repository
+            4. Delete the upload request
+
+        Args:
+            filename (str)
+                Path to a local file to upload.
+
+            relative_url (str)
+                Path that should be used in remote repository
+
+                if omitted, filename will be used.
+        Returns:
+            Future[:class:`~pubtools.pulplib.Task`]
+                A future which is resolved when publish succeeds.
+
+                The future contains the task to delete upload request
+
+        Raises:
+            DetachedException
+                If this instance is not attached to a Pulp client.
+        """
+        if not self._client:
+            raise DetachedException()
+
+        path, name = os.path.split(filename.rstrip("/"))
+        if not relative_url:
+            relative_url = path
```

Comment:
I think this is not right, looks like you've set relative_url to only the dirname and omitted the basename.

It should allow using both.  The idea is that you might have a file named locally something like tmp.json but when you import it into the repo, you should be able to import it as e.g. "some/directory/my-file.json".  Both the directory and file names don't need to be same on Pulp as they are locally.